tests/conllu_test.py

```diff
@@ -13,32 +15,52 @@ def test_swe(self):
         swe = conllu.ID(3)
         self.assertEqual(len(swe), 1)
         self.assertFalse(swe.is_mwe)
+        self.assertFalse(swe.is_decimal)
         self.assertEqual(str(swe), "3")
 
-    def test_swe_roundtrip(self):
-        swe = conllu.ID(3)
-        self.assertEqual(swe, conllu.ID.parse_from_string(str(swe)))
+    @parameterized.expand(
+        [
+            conllu.ID(3),
+            conllu.ID(2, upper=3),
+            conllu.ID(2, decimal=1),
+        ],
+    )
+    def test_roundtrip(self, the_id: conllu.ID):
+        self.assertEqual(the_id, conllu.ID.parse_from_string(str(the_id)))
 
     def test_mwe(self):
-        mwe = conllu.ID(2, 3)
+        mwe = conllu.ID(2, upper=3)
         self.assertEqual(len(mwe), 2)
         self.assertTrue(mwe.is_mwe)
+        self.assertFalse(mwe.is_decimal)
         self.assertEqual(str(mwe), "2-3")
 
-    def test_mwe_roundtrip(self):
-        mwe = conllu.ID(2, 3)
-        self.assertEqual(mwe, conllu.ID.parse_from_string(str(mwe)))
-
     def test_mwe_non_equivalence(self):
-        mwe1 = conllu.ID(2, 3)
-        mwe2 = conllu.ID(3, 4)
+        mwe1 = conllu.ID(2, upper=3)
+        mwe2 = conllu.ID(3, upper=4)
         self.assertNotEqual(mwe1, mwe2)
 
     def test_mwe_swe_non_equivalence(self):
-        mwe = conllu.ID(2, 3)
+        mwe = conllu.ID(2, upper=3)
         swe = conllu.ID(2)
         self.assertNotEqual(mwe, swe)
 
+    def test_decimal(self):
+        decimal = conllu.ID(2, decimal=1)
+        self.assertFalse(decimal.is_mwe)
+        self.assertTrue(decimal.is_decimal)
+        self.assertEqual(str(decimal), "2.1")
+
+    def test_decimal_non_equivalence(self):
+        decimal1 = conllu.ID(2, decimal=1)
+        decimal2 = conllu.ID(2, decimal=2)
+        self.assertNotEqual(decimal1, decimal2)
+
+    def test_decimal_swe_nonequivalence(self):
+        decimal = conllu.ID(2, decimal=1)
+        swe = conllu.ID(2)
```

Comment:
Question,

Would
`decimal = conllu.ID(2, decimal=0)`
ever happen? and if so, is that equal to
`swe = conllu.ID(2)`?

Reply:
Former raises an error, actually, because the spec says it's not allowed.

Internally the two are represented the same though! Because that spec detail means that I can do so. Clever.